src/common/snippets/src/lowered/shape_inference/shape_infer_instances.cpp

```diff
@@ -0,0 +1,179 @@
+                success &= broadcast_merge_dim(dims[i], dsti, srci);
+            }
+            dst = std::move(dims);
+            return success;
+        }
+        case ov::op::AutoBroadcastType::PDPD: {
+            int64_t axis = autob.m_axis;
+            if (src_rank > dst_rank || axis < -1)
+                return false;
+
+            axis = (axis == -1) ? (dst_rank - src_rank) : axis;
+            if (src_rank + axis > dst_rank)
+                return false;
+
+            bool success = true;
+            for (int64_t i = 0; i < src_rank; ++i) {
+                if (dst[axis + i] != IShapeInferSnippets::DYNAMIC_DIMENSION &&
+                    src[i] != IShapeInferSnippets::DYNAMIC_DIMENSION) {
+                    if (src[i] > dst[axis + i])
+                        return false;
+                }
+                success &= broadcast_merge_dim(dst[axis + i], dst[axis + i], src[i]);
+            }
+            return success;
+        }
+        default:
+            OPENVINO_THROW("Unsupported auto broadcast type: ", autob.m_type);
+    }
+    return false;
+}
+bool merge_into(IShapeInferSnippets::VectorDims& dst, const IShapeInferSnippets::VectorDims& src) {
+    auto merge_dim = [](size_t& dst, const size_t& d1, const size_t& d2) {
+        if (d1 == d2 || d1 == IShapeInferSnippets::DYNAMIC_DIMENSION) {
+            dst = d2;
+        } else if (d2 == IShapeInferSnippets::DYNAMIC_DIMENSION) {
+            dst = d1;
+        } else {
+            return false;
+        }
+        return true;
+    };
+    if (dst.size() != src.size())
+        return false;
+
+    bool success = true;
+    for (size_t i = 0; i < dst.size(); i++)
+        success &= merge_dim(dst[i], dst[i], src[i]);
+    return success;
+}
+} // namespace
+
+
+IShapeInferSnippets::Result
+entryNumpyBroadcasting::infer(const std::vector<std::reference_wrapper<const VectorDims>>& input_shapes) {
+        size_t max_rank = 0;
+        size_t max_rank_idx = 0;
+        for (size_t i = 0; i < input_shapes.size(); ++i) {
+            auto item_rank = input_shapes[i].get().size();
+            if (item_rank > max_rank) {
+                max_rank = item_rank;
+                max_rank_idx = i;
+            }
+        }
+        auto output_shape = input_shapes[max_rank_idx].get();
+        // use NUMPY broadcast rule
+        for (size_t i = 0; i < input_shapes.size(); i++) {
+            if (i == max_rank_idx)
+                continue;
+
+            auto& input_shape = input_shapes[i].get();
+            const size_t offset = max_rank - input_shape.size();
+            for (size_t j = 0; j < input_shape.size(); ++j) {
+                if (input_shape[j] != output_shape[offset + j]) {
+                    if (output_shape[offset + j] == 1) {
+                        output_shape[offset + j] = input_shape[j];
+                    } else if (input_shape[j] != 1) {
+                        OPENVINO_THROW(input_shapes.size() == 1, "Got not broadcastable input shapes");
```

Comment:
if the dims are not equal then one of them must be `1`. We hit this `OPENVINO_THROW` when the dims are not equal and both of them not `1`, so this exception seems legit to me 🙂 
I however agree that `input_shapes.size() == 1` is a bit confusing. More importantly, this function was written before I implemented proper nGraph-like broadcasting routines, so I think it's time to deprecate it and use `broadcast_merge_into` instead. Please take a look

Reply:
I meant that before you went to that loop only if there are several `input_shapes` (not one). And then you checked `input_shapes.size() == 1` in the loop. For sure in that loop `input_shapes.size() > 1` because you wouldn't have entered the loop body otherwise 😄 (it's like a dead code and specific analysis tools highlight these code parts). But seems like you removed that code (I cannot find it at least). So it's not actual now